### ISSUE TITLE
Remove dependency on crossovers (deprecated)

### ISSUE THREAD

**Issue by kahunamoore** [rbrush/clara-rules#53]
I read recently that Chas has deprecated the crossovers support in the lein plugin.

The suggested technique is to use cljx in order to support clojurescript.

I can make the mods but would want to do so against a fairly stable version/branch. Are there any fairly large refactorings in the works? Do you have any suggested approaches to converting to cljx?

**Comment by levand** [comment]
Just wanted to leave a note here that feature expressions are coming to Clojure, I would hold off on cljx support in favor of feature expressions built into the core of the language.

**Comment by kahunamoore** [comment]
I have been working on a pull request that converts Clara to use feature expressions... it isn't too hard. Also, recent changes in the cljs compiler has made ":require-macros true" unnecessary (AFAIK) so I was just going to drop those from the code entirely.

Luke - you are probably much better at Clojure/ClojureScript than I am. If you would rather do this let me know. Otherwise, I'll see if I can wrap up my changes soon.

Ryan - would you want to put them into an alpha branch for wider testing/development?

Thoughts?

**Comment by levand** [comment]
You go ahead, I've got plenty to do.

I'd wait to push a feature-expression branch until Clojure releases an
alpha that includes read conditionals, to avoid any last minute flux.

My sources tell me that should be soon, though.

**Comment by kahunamoore** [comment]
Great to hear... Thanks for the update!

Alan
-- 
"Whatever you can do, or dream you can do, begin it. Boldness has genius, power, and magic in it. Begin it now." - Goethe

**Comment by rbrush** [comment]
+1 for waiting for feature expressions and adopting them when available. (I'm really excited about feature expressions, by the way.).

We should keep these changes in their own feature_expression branch (or something like that) for now. Once Clojure 1.7.0 final is released we can merge them into master and upgrade Clara's dependency on Clojure itself.

**Comment by kahunamoore** [comment]
My original mods were to an old version of Clara and was based on the original "feature expressions" documentation:

http://dev.clojure.org/display/design/Feature+Expressions

... and used the older cljx syntax.

However, it looks like reader conditionals is where the action is at:

http://dev.clojure.org/display/design/Reader+Conditionals

Nice work Luke!

Looking at the latest Clara code it seems there isn't a lot that needs doing except removing old files, use RC to un-duplicate stuff, adjust some project "paths", etc. I'll let my old mods be overcome by bit-rot... and start again fresh :-)

It seems Clara doesn't use any of the ;*CLJSBUILD-MACRO-FILE*; or ;*CLJSBUILD-REMOVE*; crossover magic... so there is none of that to deal with. It seems you chose the crossovers Option #1 - Keep macros in a separate file approach.

**Comment by gitcollective** [comment]
Working on this anew, have a few questions:

1. The Rulebase record is defined in two places and one is missing an element the other definition contains. Before I go messing with this I just want to know the reason behind the location and differences in definition.
1. In merging code back into a common (.cljc) source file, would you prefer that such files live in the main/clojure directory tree or in a new peer to main/clojure, main/clojurescript? What should it be called? main/common (or main/cljc or?) Do you have a preference for where things end up?
1. Is it a goal or non-goal to get everything to live and compile in only .clj and .cljc files?
1. I was thinking of starting with only making the minimal changes needed to get rid of crossovers? I'm thinking I could just merge common code into .cljc files with clj on top, cljs on the bottom each with the appropriate reader conditional. Or, do you think going all the way and merging the reader conditionals throughout the codebase in one refactoring?
1. Are there any large changes coming from your end? I've already created a local 'reader conditionals' branch but I can fetch/rebase as needed.

I'm sure I'll have more questions and any suggestions you may have on the best way to go about this is... I'd love to know. Thanks!

**Comment by kahunamoore** [comment]
Uh, yeah. Sorry for the nonsequitur... that was me logged in under another account.

**Comment by kahunamoore** [comment]
There is a good write up of three use cases for building portable libraries in the JIRA for the reader conditionals:

http://dev.clojure.org/display/design/Reader+Conditionals#ReaderConditionals-BuildingPortableLibraries

Which of the three appeals to your goals for Clara?

**Comment by rbrush** [comment]
Hey Alan, thanks for digging into this! Comments inline...

>The Rulebase record is defined in two places and one is missing an element the other definition contains. Before I go messing with this I just want to know the reason behind the location and differences in definition.

This was almost certainly an oversight on my part. It looks like when porting to ClojureScript I split productions into rules and queries in the CLJS rule base, but didn't in CLJ. I don't think this is accessed so it won't have an impact, so I'd go with the CLJ structure since it is probably better tested.

>In merging code back into a common (.cljc) source file, would you prefer that such files live in the main/clojure directory tree or in a new peer to main/clojure, main/clojurescript? What should it be called? main/common (or main/cljc or?) Do you have a preference for where things end up?

I think src/main/clojure makes the most sense. Since the standard Clojure compiler understands .cljc files moving forward, we should just treat them as Clojure in our directory structure.

>Is it a goal or non-goal to get everything to live and compile in only .clj and .cljc files?

A nice to have, but I don't think we have to get there right away. It might be easiest just to move over one namespace at a time, ensure everything works, and then move onto the next. As we keep iterating we may eventually get everything in .clj or .cljc, but that doesn't have to be in early versions of this.

>I was thinking of starting with only making the minimal changes needed to get rid of crossovers? I'm thinking I could just merge common code into .cljc files with clj on top, cljs on the bottom each with the appropriate reader conditional. Or, do you think going all the way and merging the reader conditionals throughout the codebase in one refactoring?

This sounds like a good approach to me, at least to get started. Once everything is in cljc we can always go in and refactor the code in a subsequent pass. But getting things working in cljc with minimal refactoring is a big step in that direction, and much less prone to error than trying to do everything in one go.  

>Are there any large changes coming from your end? I've already created a local 'reader conditionals' branch but I can fetch/rebase as needed.

Nothing major planned for the next couple weeks! I might do a little work on getting agenda group/focus behavior going, but this can easily be merged with your changes. 

I would ask that you do this work on a branch and we can get everything working there. Once Clojure 1.7 goes live we'll switch Clara master to it and move the version to 0.9.0-SNAPSHOT.

>I'm sure I'll have more questions and any suggestions you may have on the best way to go about this is... I'd love to know. Thanks!

No problem, and thank you! I think the approach you propose of getting things working with a small-as-possible refactoring pass is a great idea, and then we can build on it from there. I'm pretty excited to get this in.

**Comment by kahunamoore** [comment]
Thanks for the feedback, I'll proceed in measured steps on my local branch. I forked Clara on github and cloned that locally before branching. I have remotes setup for both so I can rebase/fetch upstream changes you might be making.

The main trouble I have been running into are related to dependent library versions. I'm trying to bring the clojurescript and build plugin dependencies up to date but it feels like DLL hell all over again... I thought semantic versioning and explicit version numbers would make things easier. :-\ If you have any suggestions on this front please let me know. My current issue seems to be the prismatic macro sm/defn...

**Comment by mrrodriguez** [comment]
Not sure if this helpful. But Clara recently upgraded to schema 0.4.0 at https://github.com/rbrush/clara-rules/commit/ba92200a0bb3a0df6389aedbdd52e42498314de7. 

Unfortunately I think there will be some dependency pain around the newer versions of schema due to their breaking changes introduced https://github.com/Prismatic/schema/blob/master/CHANGELOG.md#040. 

I'm guessing this is going to make it tricky to have projects using libraries using incompatible versions of schema. Perhaps the libraries you are conflicting with have a schema version increase in their newer versions. 

Try out the 'lein deps :tree' perhaps for insight. 

(Sorry if this is all information you already know and is not useful. ;)

**Comment by kahunamoore** [comment]
Mike - thanks for the update, you probably saved me two days of hair pulling. I looked at the deps tree but didn't know at the time that schema had a breaking change. I'll make sure they are in alignment.

I've tried to use the lein-ancient plugin but always seemed to have trouble getting things to compile afterwards. I've been thinking about writing a lein-consensus plugin that scans github for clojure project.clj files, takes note of the combinations used in any given project with builds/tests that are "passing" (very important.) Next, it would correlate that information with what your project's set of dependencies are and figure out the highest version numbers of each library that are known to work together, within a user supplied confidence factor perhaps. There could be some heuristics one might apply as well to find the best set... sounds like a job for clara, core.logic or, depending on the complexity/derived state required, even miniKanren :-)

**Comment by kahunamoore** [comment]
Update - I made all the needed changes but it doesn't compile. Working through the errors now. However, my daughter is getting married next weekend so this might get delayed until the following week before I have anything that compiles. I will probably need some assistance working through the errors so expect questions starting 5/18. Let me know if this doesn't work for you.

**Comment by kahunamoore** [comment]
Ok, married off my daughter so I'm back on this. Ryan, I took a stab at this but got stuck. Before proceeding further I want to push my changes to my github fork so you can review the general approach and give feedback.

How does this sound?

**Comment by kahunamoore** [comment]
Here is a link to the branch:

https://github.com/kahunamoore/clara-rules/tree/reader-conditionals

Let me know if this is how you would like to see the code re-organized and anything else that you want to be different.

Also, looks like this branch is a few commits behind master. I'll be sure to merge those into this branch before proceeding.

**Comment by kahunamoore** [comment]
General comments: the approach taken was to simply merge the code back into common files and change the extension to .cljc where that made sense at the time. I did not try to resolve the differences between functions/macros with the same name - instead I just placed them next to each other with clj and cljs reader conditionals to choose between them. This is a large grained sort of merging of the code.

I suspect that you will want there to be only one definition of each defn/macro with more fine grained clj/cljs reader conditionals for just the specific differences between platforms.

**Comment by rbrush** [comment]
Hey Alan, thanks for working on this!

I definitely agree with the approach of just getting an initial pass of
this going in cljc, copying the content like you did.

Some good news: I checked out your branch and noticed that it was still
using Clojure 1.6 in the project.clj, so I switched it
to [org.clojure/clojure "1.7.0-RC1"] and it worked! All tests pass both in
CLJ and CLJS, although I haven't tried using the produced artifact in
anything yet.

So I think we're on our way here...next up is to probably start refactoring
the copied functions into a common base...and frequently re-running the
tests as we go so we quickly catch anything that broke.

**Comment by kahunamoore** [comment]
OMG! Seriously?!... I didn't expect it to actually work. Are you sure your build didn't find an older version of the library?

FYI: There are some random characters in one of the files so I'm not sure how it would compile. see: https://github.com/kahunamoore/clara-rules/blob/reader-conditionals/src/main/clojure/clara/rules.cljc#L178

Not sure how those characters got in there and only noticed them after pushing to the branch. I'll double check the compile with those characters removed.

Also, I didn't get rid of the rules.clj file so maybe the compile/tests are using it instead of rules.cljc. TBD.

I didn't check in the project.clj because I had hacked it up pretty badly when I had issues with the schema library version conflicts, clojure script versions, etc.

How should we proceed? I can a) get it to compile, b) do the common base refactoring you mentioned and then c) send a pull request, if that is your preference. If so, I will need some advice on some of the platform code differences and the best way to resolve them. I guess we'll just have to take those one by one.

Otherwise, let me know how you want to go forward.

**Comment by rbrush** [comment]
Yes, all I did was change the Clojure version and your branch worked...but you're right that for at least some of the namespaces it's using the existing CLJ and CLJS files. We'll probably need to upgrade the CLJS version as well to support reader conditionals there.

If you want to get it to compile and the tests passing with some basic refactoring and create a pull request, that sounds like a great way to move forward. Also please bump the version to 0.9.0-SNAPSHOT as well so we keep the 1.7 stuff there.

Thanks!

**Comment by lprefontaine** [comment]
Alan,

I can help here, I can dedicate from June 14th at least two weeks solely on this.
We are already using 1.7 and conditionals and rolling on the latest Clojure/Clojurescript stuff.

I would also write tests, a lot of them, to make sure this stuff remains interoperable between clojure
and clojurescript. I have a significant interest in this effort schedule wise for our own stuff here.

Let me know how you want to coordinate our efforts.
Luc

**Comment by kahunamoore** [comment]
Absolutely, thanks for the offer! I should have this functional by the time you are available. In the meantime feel free to suggest changes to the code - I'm still newish to Clojure as it is not part of my day job. :-(

**Comment by lprefontaine** [comment]
Hi,

finally started to work on this. I cloned your repo Alan.
Got this to load in CCW. Moved to 0.9.0-SNAPSHOT.
I upgraded dependencies except Dorothy which I understand is experimental. 

I leave for Paris this weekend and will be back next Tuesday. Alan, you can expect a pull
request by the end of next week. I'll also back this up with my private git repo in the cloud.

Should we create a separate issue at this point ?
Any other comments ?

**Comment by lprefontaine** [comment]
Another thing, I intend to be quite systematic here, I will try to consolidate as much as 
possible Clojure and ClojureScript in a single source base. Aside from the obvious JVM
dependencies.

First pass, I'll use a dumb approach, I'll start with the leaf in the dependency graph
and move upward. That should help me finding obvious cross-platform show stoppers.

Second pass, see what can be done to increase the common code portion.

Comments ?

**Comment by kahunamoore** [comment]
I have made changes since the version on github. The current sticking point is part of a large deftype expression where the only difference between cljs and clj is a meta data type hint.

I thought it might be a bug in the reader conditional implementation but when I went to create a minimum reproducible case/project it didn't fail.

The "workaround" would be to duplicate that very large typedef which will be my next step if I can't prove it is a problem with the compiler.

**Comment by lprefontaine** [comment]
I am almost done. Meta data it seems cannot be wrapped with conditionals.
I made a duplicate of this type def for the moment.

I opened a new issue instead of adding to this one. Can't remember the #, just check for clojure 1.7/reader conditionals/...

**Comment by rbrush** [comment]
Sounds great! I'm traveling today but am looking forward to seeing this.
The strategy discussed so far makes sense to me, including the simple first
pass and refactoring.

**Comment by kahunamoore** [comment]
Yes, that is what I concluded. I searched the docs but couldn't find a definitive answer one way or the other. I was trying to figure out another way to do it w/o copying but I guess that will have to do for now.

Maybe that meta data can be applied *within* the called function or in the caller instead of in the args vector... TBD. I'm not that familiar with the JVM side of Clojure, I spend most of my time prototyping in cljs. I will be working on the server side soon so I'd like to understand those details better.

Glad you are making progress and look forward to getting this finally resolved!
